File: frontend/src/lib/types/podcast.ts
// Podcast generation types

export type PodcastStyle = 
  | "conversational" 
  | "interview" 
  | "educational" 
  | "debate" 
  | "storytelling";

export type VoiceName = 
  | "Kore"    // Male voice
  | "Puck"    // Female voice
  | "Charon"  // Deep male voice
  | "Fenrir"  // Energetic male voice
  | "Aoede"   // Warm female voice
  | "Leda";   // Professional female voice

export interface SpeakerConfig {
  name: string;
  voice: VoiceName;
  role: string;
}

export interface PodcastRequest {
  sources: Array<
    | { type: "file"; file_id: string }
    | { type: "url"; url: string }
    | { type: "text"; content: string }
  >;
  style: PodcastStyle;
  provider: "gemini" | "google" | "openai" | "anthropic";
  model: string;
  speakers: SpeakerConfig[];
  duration_minutes: number;
}

// SSE Event types
export interface PodcastProgressEvent {
  type: "progress";
  stage: string;
  percent: number;
  message?: string;
}

export interface PodcastCompleteEvent {
  type: "complete";
  title: string;
  description: string;
  audio_base64: string;
  script: Array<{ speaker: string; text: string }>;
  duration_seconds: number;
}

export interface PodcastErrorEvent {
  type: "error";
  message: string;
  code?: string;
}

export type PodcastEvent = PodcastProgressEvent | PodcastCompleteEvent | PodcastErrorEvent;

// Type guards
export function isPodcastProgressEvent(event: PodcastEvent): event is PodcastProgressEvent {
  return event.type === "progress";
}

export function isPodcastCompleteEvent(event: PodcastEvent): event is PodcastCompleteEvent {
  return event.type === "complete";
}

export function isPodcastErrorEvent(event: PodcastEvent): event is PodcastErrorEvent {
  return event.type === "error";
}

// Voice options for UI
export const VOICE_OPTIONS: Array<{ value: VoiceName; label: string; description: string }> = [
  { value: "Kore", label: "Kore", description: "Male voice" },
  { value: "Puck", label: "Puck", description: "Female voice" },
  { value: "Charon", label: "Charon", description: "Deep male voice" },
  { value: "Fenrir", label: "Fenrir", description: "Energetic male voice" },
  { value: "Aoede", label: "Aoede", description: "Warm female voice" },
  { value: "Leda", label: "Leda", description: "Professional female voice" },
];

// Style options for UI
export const PODCAST_STYLE_OPTIONS: Array<{
  value: PodcastStyle;
  label: string;
  icon: string;
  description: string;
}> = [
  {
    value: "conversational",
    label: "Casual Chat",
    icon: "💬",
    description: "Friendly discussion between hosts",
  },
  {
    value: "interview",
    label: "Interview",
    icon: "🎤",
    description: "Host interviewing an expert",
  },
  {
    value: "educational",
    label: "Educational",
    icon: "📚",
    description: "Teaching/explaining the topic",
  },
  {
    value: "debate",
    label: "Debate",
    icon: "⚖️",
    description: "Two opposing viewpoints",
  },
  {
    value: "storytelling",
    label: "Storytelling",
    icon: "📖",
    description: "Narrative-driven discussion",
  },
];

// Default speakers
export const DEFAULT_SPEAKERS: SpeakerConfig[] = [
  { name: "Alex", voice: "Kore", role: "host" },
  { name: "Sam", voice: "Puck", role: "co-host" },
];

File: backend/doc_generator/infrastructure/api/schemas/podcast.py
"""Pydantic request and response models for podcast generation."""

from enum import Enum
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from .requests import Provider, SourceItem


class PodcastStyle(str, Enum):
    """Podcast generation styles."""

    CONVERSATIONAL = "conversational"  # Casual chat between hosts
    INTERVIEW = "interview"  # Host interviewing an expert
    EDUCATIONAL = "educational"  # Teaching/explaining content
    DEBATE = "debate"  # Two opposing viewpoints
    STORYTELLING = "storytelling"  # Narrative style


class VoiceName(str, Enum):
    """Available voice names for TTS."""

    KORE = "Kore"  # Male voice
    PUCK = "Puck"  # Female voice
    CHARON = "Charon"  # Deep male voice
    FENRIR = "Fenrir"  # Energetic male voice
    AOEDE = "Aoede"  # Warm female voice
    LEDA = "Leda"  # Professional female voice


class SpeakerConfig(BaseModel):
    """Configuration for a podcast speaker."""

    name: str = Field(description="Speaker name used in the script")
    voice: VoiceName = Field(description="TTS voice to use for this speaker")
    role: str = Field(default="host", description="Speaker role (host, guest, expert)")


class PodcastRequest(BaseModel):
    """Request model for podcast generation.

    Example:
        {
            "sources": [
                {"type": "file", "file_id": "f_abc123"},
                {"type": "url", "url": "https://example.com/article"},
                {"type": "text", "content": "Some text content..."}
            ],
            "style": "conversational",
            "provider": "gemini",
            "model": "gemini-2.5-flash",
            "speakers": [
                {"name": "Alex", "voice": "Kore", "role": "host"},
                {"name": "Sam", "voice": "Puck", "role": "co-host"}
            ],
            "duration_minutes": 5
        }
    """

    sources: list[SourceItem] = Field(
        description="List of sources (file, url, or text)",
        min_length=1,
    )
    style: PodcastStyle = PodcastStyle.CONVERSATIONAL
    provider: Provider = Provider.GEMINI
    model: str = "gemini-2.5-flash"
    speakers: list[SpeakerConfig] = Field(
        default_factory=lambda: [
            SpeakerConfig(name="Alex", voice=VoiceName.KORE, role="host"),
            SpeakerConfig(name="Sam", voice=VoiceName.PUCK, role="co-host"),
        ],
        min_length=2,
        max_length=4,
        description="Speaker configurations (2-4 speakers)",
    )
    duration_minutes: int = Field(
        default=3,
        ge=1,
        le=10,
        description="Target podcast duration in minutes",
    )

    model_config = ConfigDict(
        json_schema_extra={
            "examples": [
                {
                    "sources": [
                        {"type": "url", "url": "https://example.com/article"},
                        {"type": "text", "content": "Additional context"},
                    ],
                    "style": "conversational",
                    "provider": "gemini",
                    "model": "gemini-2.5-flash",
                    "speakers": [
                        {"name": "Alex", "voice": "Kore", "role": "host"},
                        {"name": "Sam", "voice": "Puck", "role": "co-host"},
                    ],
                    "duration_minutes": 5,
                }
            ]
        }
    )


class PodcastScript(BaseModel):
    """Generated podcast script with dialogue."""

    title: str
    description: str
    dialogue: list[dict]  # List of {speaker: str, text: str}


# SSE Event Models


class PodcastProgressEvent(BaseModel):
    """Progress event during podcast generation."""

    type: Literal["progress"] = "progress"
    stage: str  # extracting, scripting, synthesizing
    percent: float = Field(ge=0, le=100)
    message: str | None = None


class PodcastCompleteEvent(BaseModel):
    """Completion event with podcast audio data."""

    type: Literal["complete"] = "complete"
    title: str
    description: str
    audio_base64: str  # Base64-encoded WAV audio
    script: list[dict]  # The generated dialogue script
    duration_seconds: float


class PodcastErrorEvent(BaseModel):
    """Error event during podcast generation."""

    type: Literal["error"] = "error"
    message: str
    code: str | None = None
